### ISSUE TITLE
Make it possible to initialize array from dictionary values in Python 3

### ISSUE THREAD

**Issue by astrofrog** [numpy/numpy#5718]
At the moment arrays can't be initialized from dictionary values in Python 3:

```
In [1]: import numpy as np

In [2]: d = {'a':1,'b':2,'c':3}

In [3]: np.array(d.values())
Out[3]: array(dict_values([2, 1, 3]), dtype=object)
```

and one has to instead explicitly convert to a list first:

```
In [4]: np.array(list(d.values()))
Out[4]: array([2, 1, 3])
```

It might be worth adding a special case in the initializer for the array to automatically do this conversion, since ``d.values()`` is a valid iterator?

**Comment by pv** [comment]
That seems like a behavior change that would affect all iterators. I'm
not sure it's a good idea to add workarounds for changes in semantics in
Python 3.

**Comment by jaimefrio** [comment]
The right approach for this would be:

```
>>> import numpy as np
>>> d = {'a':1,'b':2,'c':3}
>>> np.fromiter(d.itervalues(), dtype=np.int)  # I'm on Python 2.7
array([1, 3, 2])
```

Unfortunately, it does not work for `object` dtype:

```
>>> np.fromiter(d.itervalues(), dtype=object)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: cannot create object arrays from iterator
```

If that were fixed (I think it has to do with an overzealous programmer worrying too much about reference counting going wrong), would this solution fit your needs?

**Comment by astrofrog** [comment]
@jaimefrio - I personally am fine with this workaround, but more generally this is something that beginners will stumble against. Since this works in Python 2, I don't see why conceptually it would be useful to have a different behavior in Python 3 (and it's going to break existing user code). I would **strongly** advocate trying to maintain backward compatibility here with Python 2 if at all possible even if it means adding a special case for dictvalues and dictkeys, since it's going to be a pretty common case.

**Comment by jaimefrio** [comment]
I don't think that `array` has a way of knowing that what it received is a `dict.keys()` or `dict.values()` iterator. Any change would require accepting generic iterators, which is not an easy task, e.g. `np.fromiter` requires an explicit `dtype`, while `np.array` can figure it out from the input. So I don't really see an easy way of making this happen...

**Comment by njsmith** [comment]
No, in python 3, dict.keys and dict.values (and also dict.items for that
matter) don't return iterators, they return special magical view classes.
Which are iterable, of course, but they also have specific documented types.

I too am dubious about np.array accepting arbitrary iterables (because of
memory overhead and the difficulty of even recognizing an iterable when you
see one), but I'm +0.7 on teaching np.array to treat the dict views the
same way we treat lists.

**Comment by eric-wieser** [comment]
What is actually going on here?

This works fine:

```
class A(set):
    def __getitem__(self, index):
        if index > 3:
            raise IndexError
        return index

    def __len__(self): return 4

np.array(A())  # array([1, 2, 3, 4])
```

So why does `np.array(dict_keys)` not?

**Comment by KelSolaar** [comment]
Hi,

I'm finding every so and then having to cast to `list` first and it is mildly annoying having to do that. Any resolution for this one?

Cheers,

Thomas

**Comment by bje-** [comment]
On a related matter, is it possible to do something like this elegantly without using a loop?

```
>>> d = {0: 12, 3: 20, 5: 24}
>>> z = np.init_from_dict(d, missing=0)
>>> z
array([12.,  0.,  0., 20.,  0., 24.])
```

**Comment by obackhouse** [comment]
> On a related matter, is it possible to do something like this elegantly without using a loop?
> 
> ```
> >>> d = {0: 12, 3: 20, 5: 24}
> >>> z = np.init_from_dict(d, missing=0)
> >>> z
> array([12.,  0.,  0., 20.,  0., 24.])
> ```

```
z = np.zeros(max(d.keys())+1)
z[list(d.keys())] = list(d.values())
```
